File: libraries/react-native/ios/HybridMeasurement.swift
import Foundation
import NitroModules

final class HybridMeasurement: HybridMeasurementSpec {
    private struct RemoteConfiguration: Decodable {
        struct Rule: Decodable {
            let coarseValue: ConversionValueCoarse?
            let eventName: String
            let fineValue: Int
            let lockWindow: Bool?
            let minimumCount: Int
            let window: Int
        }
        let conversionRules: [Rule]
        let schemaVersion: Int
    }
    private var readiness = "uninitialized"
    private var consentRevision = 0.0
    private var configurationRevision = 0.0
    private var listeners: [String: (MeasurementBridgeEvent) -> Void] = [:]
    private var publishableKey: String?
    private var ingestOrigin: URL?
    private lazy var store: MeasurementStore = {
        do { return try MeasurementStore() }
        catch { fatalError("MEASUREMENT_STORE_INITIALIZATION_FAILED") }
    }()
    private lazy var conversionEngine = ConversionValueEngine(
        adapter: SystemConversionValuePlatformAdapter(),
        persistRules: { _, _ in true },
        evidenceSink: { _ in }
    )

    private func snapshot() throws -> MeasurementStateBridge {
        let state = try store.snapshot()
        return MeasurementStateBridge(
            installationId: state.installationId,
            firstOpenedAt: state.firstOpenedAt,
            installationSequence: Double(state.sequence),
            readiness: readiness,
            currentSessionId: nil,
            currentSessionSequence: nil,
            consentRevision: consentRevision,
            configurationRevision: configurationRevision,
            outboxCritical: Double(state.counts["critical"] ?? 0),
            outboxHigh: Double(state.counts["high"] ?? 0),
            outboxNormal: Double(state.counts["normal"] ?? 0),
            outboxLow: Double(state.counts["low"] ?? 0),
            oldestRecordAgeMs: state.oldestQueuedAtMs.map {
                Double(max(0, Int64(Date().timeIntervalSince1970 * 1_000) - $0))
            }
        )
    }

    func initialize(publishableKey: String, configuration: MeasurementInitializeConfiguration) throws -> Promise<MeasurementStateBridge> {
        guard !publishableKey.isEmpty else { throw RuntimeError.error(withMessage: "INVALID_PUBLISHABLE_KEY") }
        return Promise.async {
            _ = try self.store.snapshot()
            self.publishableKey = publishableKey
            self.ingestOrigin = URL(string: configuration.ingestUrl)
            self.configurationRevision += 1
            self.readiness = "sdkReady"
            return try self.snapshot()
        }
    }

    func enqueue(command: MeasurementCommand) throws -> Promise<MeasurementCommandResult> {
        let publicPayload = Data(bytes: command.publicPayload.data, count: command.publicPayload.size)
        return Promise.async {
            let recordId = command.commandId
            let sequence = try self.store.enqueue(
                recordId: recordId,
                recordType: command.recordType,
                occurredAt: command.occurredAt,
                priority: String(describing: command.priority),
                source: String(describing: command.source),
                publicPayload: String(decoding: publicPayload, as: UTF8.self),
                protectedPayloadRef: command.protectedEvidenceRef
            )
            if let consent = command.consent { self.consentRevision = consent.revision }
            return MeasurementCommandResult(
                accepted: true,
                recordId: recordId,
                installationSequence: Double(sequence),
                error: nil
            )
        }
    }

    func flush() throws -> Promise<MeasurementFlushBridgeResult> {
        Promise.async {
            guard let key = self.publishableKey, let origin = self.ingestOrigin else {
                let count = try self.store.peekEligible(limit: Int.max).count
                return MeasurementFlushBridgeResult(accepted: 0, scheduled: Double(count), quarantined: 0, policyBlocked: 0)
            }
            let result = await MeasurementDelivery(
                store: self.store,
                publishableKey: key,
                ingestOrigin: origin
            ).flush()
            return MeasurementFlushBridgeResult(
                accepted: Double(result.accepted),
                scheduled: Double(result.scheduled),
                quarantined: Double(result.quarantined),
                policyBlocked: Double(result.policyBlocked)
            )
        }
    }

    func getInstallationId() throws -> Promise<String> {
        Promise.async { try self.store.snapshot().installationId }
    }

    func getState() throws -> Promise<MeasurementStateBridge> { Promise.async { try self.snapshot() } }

    func subscribe(subscriptionId: String, listener: @escaping (MeasurementBridgeEvent) -> Void) throws {
        listeners[subscriptionId] = listener
    }

    func unsubscribe(subscriptionId: String) throws { listeners.removeValue(forKey: subscriptionId) }

    func peekInbox(limit: Double) throws -> Promise<[MeasurementInboxEntry]> {
        Promise.async {
            try self.store.peekInbox(limit: Int(limit)).map {
                MeasurementInboxEntry(
                    id: $0.id,
                    kind: $0.kind,
                    source: $0.source,
                    appState: $0.appState,
                    receivedAt: $0.receivedAt,
                    protectedEvidenceRef: $0.protectedPayloadRef
                )
            }
        }
    }

    func acknowledgeInbox(entryId: String) throws -> Promise<Bool> {
        Promise.async { try self.store.acknowledgeInbox(id: entryId) }
    }

    func readProtectedEvidence(blobId: String) throws -> Promise<ArrayBuffer> {
        Promise.async {
            guard let evidence = try self.store.getProtectedEvidence(blobId: blobId) else {
                throw RuntimeError.error(withMessage: "PROTECTED_EVIDENCE_NOT_FOUND")
            }
            return try ArrayBuffer.copy(data: evidence.value)
        }
    }

    func putProtectedEvidence(input: MeasurementProtectedEvidenceInput) throws -> Promise<String> {
        let value = Data(bytes: input.value.data, count: input.value.size)
        return Promise.async {
            try self.store.putProtectedEvidence(
                blobId: input.blobId,
                purpose: input.purpose.stringValue,
                consentRevision: Int64(input.consentRevision),
                retentionClass: input.retentionClass.stringValue,
                value: value
            )
        }
    }

    func deleteProtectedEvidence(blobId: String) throws -> Promise<Bool> {
        Promise.async { try self.store.deleteProtectedEvidence(blobId: blobId) }
    }

    func deleteProtectedData(requestId: String) throws -> Promise<Bool> {
        Promise.async { try self.store.deleteProtectedData(requestId: requestId) }
    }

    func getMeasurementConfigurationState() throws -> Promise<MeasurementConfigurationStateBridge> {
        Promise.async {
            let state = try self.store.measurementConfigurationState()
            return MeasurementConfigurationStateBridge(
                version: Double(state.version),
                payload: try state.payload.map { try ArrayBuffer.copy(data: $0) }
            )
        }
    }

    func persistMeasurementConfigurationState(version: Double, payload: ArrayBuffer) throws -> Promise<Bool> {
        let bytes = Data(bytes: payload.data, count: payload.size)
        return Promise.async {
            try self.store.persistMeasurementConfiguration(version: Int64(version), payload: bytes)
        }
    }

    func applyMeasurementConfiguration(version: Double, payload: ArrayBuffer) throws -> Promise<Void> {
        let bytes = Data(bytes: payload.data, count: payload.size)
        return Promise.async {
            let configuration = try JSONDecoder().decode(RemoteConfiguration.self, from: bytes)
            guard configuration.schemaVersion == 1 else {
                throw RuntimeError.error(withMessage: "MEASUREMENT_CONFIGURATION_INVALID")
            }
            let rules = configuration.conversionRules.map {
                ConversionValueRule(
                    eventName: $0.eventName,
                    minimumCount: $0.minimumCount,
                    fineValue: $0.fineValue,
                    coarseValue: $0.coarseValue,
                    lockWindow: $0.lockWindow ?? false,
                    window: $0.window
                )
            }
            try self.conversionEngine.applyRules(version: Int64(version), rules: rules)
        }
    }

    func applyMeasurementStorageLimits(
        maxOutboxRecords: Double,
        maxOutboxBytes: Double,
        maxProtectedBytes: Double
    ) throws -> Promise<Void> {
        Promise.async {
            try self.store.applyStorageLimits(
                maxOutboxRecords: Int64(maxOutboxRecords),
                maxOutboxBytes: Int64(maxOutboxBytes),
                maxProtectedBytes: Int64(maxProtectedBytes)
            )
        }
    }

    func getPushRegistrationState() throws -> Promise<MeasurementConfigurationStateBridge> {
        Promise.async {
            let state = try self.store.pushRegistrationState()
            return MeasurementConfigurationStateBridge(
                version: Double(state.version),
                payload: try state.payload.map { try ArrayBuffer.copy(data: $0) }
            )
        }
    }

    func persistPushRegistrationState(payload: ArrayBuffer) throws -> Promise<Bool> {
        let bytes = Data(bytes: payload.data, count: payload.size)
        return Promise.async { try self.store.persistPushRegistration(payload: bytes) }
    }

    func clearPushRegistrationState() throws -> Promise<Bool> {
        Promise.async { try self.store.clearPushRegistration() }
    }

    func getTestDeviceState() throws -> Promise<Bool> {
        Promise.async { try self.store.testDeviceState() }
    }

    func persistTestDeviceState(enabled: Bool) throws -> Promise<Bool> {
        Promise.async { try self.store.persistTestDeviceState(enabled) }
    }

    func checkAndSetDedupe(namespace: String, key: String, expiresAtMs: Double) throws -> Promise<Bool> {
        Promise.async {
            try self.store.checkAndSetDedupe(
                namespace: namespace,
                key: key,
                expiresAtMs: Int64(expiresAtMs)
            )
        }
    }


    func hasDedupe(namespace: String, key: String) throws -> Promise<Bool> {
        Promise.async { try self.store.hasDedupe(namespace: namespace, key: key) }
    }
}

File: libraries/react-native/android/src/main/java/com/margelo/nitro/voidhash/HybridMeasurement.kt
package com.margelo.nitro.voidhash

import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.voidhash.measurement.MeasurementStore
import com.margelo.nitro.voidhash.measurement.MeasurementDelivery
import com.margelo.nitro.voidhash.measurement.GooglePlayInstallReferrerProvider
import com.margelo.nitro.voidhash.measurement.InstallReferrerCoordinator
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.nio.ByteBuffer
import org.json.JSONObject
import kotlin.coroutines.Continuation
import kotlin.coroutines.EmptyCoroutineContext
import kotlin.coroutines.startCoroutine

class HybridMeasurement : HybridMeasurementSpec() {
    private val listeners = ConcurrentHashMap<String, (MeasurementBridgeEvent) -> Unit>()
    private var readiness = "uninitialized"
    private var consentRevision = 0.0
    private var configurationRevision = 0.0
    private var publishableKey: String? = null
    private var ingestOrigin: String? = null
    private val installReferrerStarted = AtomicBoolean(false)
    private var appliedConfigurationVersion = 0L

    private val store: MeasurementStore by lazy {
        val context = NitroModules.applicationContext
            ?: throw IllegalStateException("NITRO_CONTEXT_UNAVAILABLE")
        MeasurementStore(context)
    }

    private fun snapshot(): MeasurementStateBridge {
        val state = store.snapshot()
        return MeasurementStateBridge(
            installationId = state.installationId,
            firstOpenedAt = state.firstOpenedAt,
            installationSequence = state.sequence.toDouble(),
            readiness = readiness,
            currentSessionId = null,
            currentSessionSequence = null,
            consentRevision = consentRevision,
            configurationRevision = configurationRevision,
            outboxCritical = (state.counts["critical"] ?: 0).toDouble(),
            outboxHigh = (state.counts["high"] ?: 0).toDouble(),
            outboxNormal = (state.counts["normal"] ?: 0).toDouble(),
            outboxLow = (state.counts["low"] ?: 0).toDouble(),
            oldestRecordAgeMs = state.oldestQueuedAtMs?.let { (System.currentTimeMillis() - it).coerceAtLeast(0).toDouble() },
        )
    }

    override fun initialize(publishableKey: String, configuration: MeasurementInitializeConfiguration): Promise<MeasurementStateBridge> = Promise.async {
        require(publishableKey.isNotBlank()) { "INVALID_PUBLISHABLE_KEY" }
        store.snapshot()
        this.publishableKey = publishableKey
        ingestOrigin = configuration.ingestUrl
        configurationRevision += 1
        readiness = "sdkReady"
        if (installReferrerStarted.compareAndSet(false, true)) {
            val context = NitroModules.applicationContext
                ?: throw IllegalStateException("NITRO_CONTEXT_UNAVAILABLE")
            suspend {
                InstallReferrerCoordinator(
                    store,
                    GooglePlayInstallReferrerProvider(context),
                ).collectOnce()
            }.startCoroutine(object : Continuation<Unit> {
                override val context = EmptyCoroutineContext
                override fun resumeWith(result: Result<Unit>) = Unit
            })
        }
        snapshot()
    }

    override fun enqueue(command: MeasurementCommand): Promise<MeasurementCommandResult> = Promise.async {
        val buffer = command.publicPayload.getBuffer(true)
        buffer.rewind()
        val publicPayload = ByteArray(buffer.remaining())
        buffer.get(publicPayload)
        val sequence = store.enqueue(
            recordId = command.commandId,
            recordType = command.recordType,
            occurredAt = command.occurredAt,
            priority = command.priority.name,
            source = command.source.name,
            publicPayload = publicPayload.toString(Charsets.UTF_8),
            protectedPayloadRef = command.protectedEvidenceRef,
        )
        command.consent?.let { consentRevision = it.revision }
        MeasurementCommandResult(true, command.commandId, sequence.toDouble(), null)
    }

    override fun flush(): Promise<MeasurementFlushBridgeResult> = Promise.async {
        val key = publishableKey
        val origin = ingestOrigin
        if (key == null || origin.isNullOrBlank()) {
            val scheduled = store.peekEligible(Int.MAX_VALUE).size
            MeasurementFlushBridgeResult(0.0, scheduled.toDouble(), 0.0, 0.0)
        } else {
            val result = MeasurementDelivery(store, key, origin).flush()
            MeasurementFlushBridgeResult(
                result.accepted.toDouble(),
                result.scheduled.toDouble(),
                result.quarantined.toDouble(),
                result.policyBlocked.toDouble(),
            )
        }
    }

    override fun getInstallationId(): Promise<String> = Promise.async {
        store.snapshot().installationId
    }

    override fun getState(): Promise<MeasurementStateBridge> = Promise.async { snapshot() }

    override fun subscribe(subscriptionId: String, listener: (MeasurementBridgeEvent) -> Unit) {
        listeners[subscriptionId] = listener
    }

    override fun unsubscribe(subscriptionId: String) {
        listeners.remove(subscriptionId)
    }

    override fun peekInbox(limit: Double): Promise<Array<MeasurementInboxEntry>> =
        Promise.async {
            store.peekInbox(limit.toInt()).map {
                MeasurementInboxEntry(it.id, it.kind, it.source, it.appState, it.receivedAt, it.protectedPayloadRef)
            }.toTypedArray()
        }

    override fun acknowledgeInbox(entryId: String): Promise<Boolean> = Promise.async {
        store.acknowledgeInbox(entryId)
    }

    override fun readProtectedEvidence(blobId: String): Promise<ArrayBuffer> = Promise.async {
        val evidence = store.getProtectedEvidence(blobId)
            ?: throw IllegalStateException("PROTECTED_EVIDENCE_NOT_FOUND")
        ArrayBuffer.copy(ByteBuffer.wrap(evidence.value))
    }

    override fun putProtectedEvidence(input: MeasurementProtectedEvidenceInput): Promise<String> = Promise.async {
        val buffer = input.value.getBuffer(true)
        buffer.rewind()
        val value = ByteArray(buffer.remaining())
        buffer.get(value)
        store.putProtectedEvidence(
            blobId = input.blobId,
            purpose = input.purpose.name.lowercase().replace('_', '-'),
            consentRevision = input.consentRevision.toLong(),
            retentionClass = input.retentionClass.name.lowercase().replace('_', '-'),
            value = value,
        )
    }

    override fun deleteProtectedEvidence(blobId: String): Promise<Boolean> = Promise.async {
        store.deleteProtectedEvidence(blobId)
    }

    override fun deleteProtectedData(requestId: String): Promise<Boolean> = Promise.async {
        store.deleteProtectedData(requestId)
    }

    override fun getMeasurementConfigurationState(): Promise<MeasurementConfigurationStateBridge> = Promise.async {
        val state = store.measurementConfigurationState()
        MeasurementConfigurationStateBridge(
            state.version.toDouble(),
            state.payload?.let { ArrayBuffer.copy(ByteBuffer.wrap(it)) },
        )
    }

    override fun persistMeasurementConfigurationState(version: Double, payload: ArrayBuffer): Promise<Boolean> = Promise.async {
        val buffer = payload.getBuffer(true)
        buffer.rewind()
        val bytes = ByteArray(buffer.remaining())
        buffer.get(bytes)
        store.persistMeasurementConfiguration(version.toLong(), bytes)
    }

    override fun applyMeasurementConfiguration(version: Double, payload: ArrayBuffer): Promise<Unit> = Promise.async {
        require(version.toLong() > appliedConfigurationVersion) { "MEASUREMENT_CONFIGURATION_VERSION_REPLAY" }
        val buffer = payload.getBuffer(true)
        buffer.rewind()
        val bytes = ByteArray(buffer.remaining())
        buffer.get(bytes)
        val decoded = JSONObject(bytes.toString(Charsets.UTF_8))
        require(decoded.optInt("schemaVersion") == 1) { "MEASUREMENT_CONFIGURATION_INVALID" }
        appliedConfigurationVersion = version.toLong()
    }

    override fun applyMeasurementStorageLimits(
        maxOutboxRecords: Double,
        maxOutboxBytes: Double,
        maxProtectedBytes: Double,
    ): Promise<Unit> = Promise.async {
        store.applyStorageLimits(
            maxOutboxRecords.toInt(),
            maxOutboxBytes.toLong(),
            maxProtectedBytes.toLong(),
        )
    }

    override fun getPushRegistrationState(): Promise<MeasurementConfigurationStateBridge> = Promise.async {
        val state = store.pushRegistrationState()
        MeasurementConfigurationStateBridge(
            state.version.toDouble(),
            state.payload?.let { ArrayBuffer.copy(ByteBuffer.wrap(it)) },
        )
    }

    override fun persistPushRegistrationState(payload: ArrayBuffer): Promise<Boolean> = Promise.async {
        val buffer = payload.getBuffer(true)
        buffer.rewind()
        val bytes = ByteArray(buffer.remaining())
        buffer.get(bytes)
        store.persistPushRegistration(bytes)
    }

    override fun clearPushRegistrationState(): Promise<Boolean> = Promise.async {
        store.clearPushRegistration()
    }

    override fun getTestDeviceState(): Promise<Boolean> = Promise.async { store.testDeviceState() }

    override fun persistTestDeviceState(enabled: Boolean): Promise<Boolean> = Promise.async {
        store.persistTestDeviceState(enabled)
    }

    override fun checkAndSetDedupe(namespace: String, key: String, expiresAtMs: Double): Promise<Boolean> = Promise.async {
        store.checkAndSetDedupe(namespace, key, expiresAtMs.toLong())
    }

    override fun hasDedupe(namespace: String, key: String): Promise<Boolean> = Promise.async {
        store.hasDedupe(namespace, key)
    }
}
